neps/nep-0399.md

```diff
@@ -0,0 +1,375 @@
+for the state that this `FlatState` object represents, i.e., the state that after
+block `self.block_hash` is applied.
+
+`FlatState` will be stored as a field in `Tries`. 
+
+###```ShardFlatStates```
+`ShardFlatStates` will be stored as part of `ShardTries`. Similar to how `ShardTries` is used to 
+construct new `Trie` objects given a state root and a shard id, `ShardFlatStates` is used to construct
+a new `FlatState` object given a block hash and a shard id. 
+
+```rust
+pub fn new_flat_state_for_shard(
+    &self,
+    shard_id: ShardId,
+    block_hash: Option<CryptoHash>,
+) -> FlatState
+```
+Creates a new `FlatState` to be used for performing key value lookups on the state of shard `shard_id`
+after block `block_hash` is applied.
+
+```rust
+pub fn get_flat_storage_state_for_shard(
+    &self,
+    shard_id: ShardId,
+) -> Result<FlatStorageState, FlatStorageError>
+```
+Returns the `FlatStorageState` for the shard `shard_id`. This function is needed because even though 
+`FlatStorageState` is part of `Runtime`, `Chain` also needs access to `FlatStorageState` to update flat head. 
+We will also create a function with the same in `Runtime` that calls this function to provide `Chain` to access
+to `FlatStorageState`.
+
+###```FlatStorageState```
+`FlatStorageState` is created per shard. It provides information to which blocks the flat storage 
+on the given shard currently supports and what block deltas need to be applied on top the stored
+flat state on disk to get the state of the target block. 
+
+```rust
+fn get_deltas_between_blocks(
+    &self,
+    target_block_hash: &CryptoHash,
+) -> Result<Vec<Arc<FlatStateDelta>>, FlatStorageError>
+```
+Returns the list of deltas between blocks `target_block_hash`(inclusive) and flat head(exclusive), 
+Returns an error if `target_block_hash` is not a direct descendent of the current flat head.
+This function will be used in `FlatState::get_ref`.
+
+```rust
+fn update_flat_head(&self, new_head: &CryptoHash) -> Result<(), FlatStorageError>
+```
+Updates the head of the flat storage, including updating the flat head in memory and on disk,
+update the flat state on disk to reflect the state at the new head, and gc the `FlatStateDelta`s that 
+are no longer needed from memory and from disk. 
+
+```rust
+fn add_delta(
+    &self,
+    block_hash: &CryptoHash,
+    delta: FlatStateDelta,
+) -> Result<StoreUpdate, FlatStorageError>
+```
+Adds `delta` to `FlatStorageState`, returns a `StoreUpdate` object that includes 
+
+#### Thread Safety
+We should note that the implementation of `FlatStorageState` must be thread safe because it can 
+be concurrently accessed by multiple threads. A node can process multiple blocks at the same time
+if they are on different forks. Therefore, `FlatStorageState` will be guarded by a `RwLock` so its
+access can be shared safely.
+
+```rust
+pub struct FlatStorageState(Arc<RwLock<FlatStorageStateInner>>);
+```
+
+## Drawbacks
+
```

Comment:
also flat storage would increase the disk requirements from the validators

Reply:
Mentioned that - however, `DBCol::State` is several times bigger.